htdocs/core/class/html.form.class.php

```diff
@@ -3312,55 +3312,70 @@ function formSelectShippingMethod($page, $selected='', $htmlname='shipping_metho
 	/**
 	 * Creates HTML last in cycle situation invoices selector
 	 *
-	 * @param     string  $selected   		Preselected ID
-	 * @param     int     $socid      		Company ID
+	 * @param     string	$selected		Preselected ID
+	 * @param     int		$socid			Company ID
+	 * @param     string	$htmlname		Select name
 	 *
 	 * @return    string                     HTML select
 	 */
-	function selectSituationInvoices($selected = '', $socid = 0)
+	function selectSituationInvoices($selected = null, $socid = null, $htmlname = 'situations')
 	{
 		global $langs;
 
 		$langs->load('bills');
 
-		$opt = '<option value ="" selected></option>';
-		$sql = 'SELECT rowid, facnumber, situation_cycle_ref, situation_counter, situation_final, fk_soc FROM ' . MAIN_DB_PREFIX . 'facture WHERE situation_counter>=1';
-		$sql.= ' ORDER by situation_cycle_ref, situation_counter desc';
+		$sql = 'SELECT rowid, facnumber, situation_cycle_ref, situation_counter';
+		$sql.= ' FROM ' . MAIN_DB_PREFIX . 'facture';
+		$sql.= ' WHERE entity IN ('.getEntity('invoice').')';
+		$sql.= ' AND situation_counter >= 1';
+		$sql.= " AND facnumber NOT LIKE '%PROV%'";
+		$sql.= " AND situation_final != 1";
+		if (! empty($socid)) {
+			$sql.= ' AND fk_soc = ' . $socid;
+		}
+		$sql.= ' ORDER BY situation_cycle_ref, situation_counter DESC';
+
 		$resql = $this->db->query($sql);
-		if ($resql && $this->db->num_rows($resql) > 0) {
-			// Last seen cycle
-			$ref = 0;
-			while ($obj = $this->db->fetch_object($resql)){
-				//Same company ?
-			    if ($socid == $obj->fk_soc) {
+
+		if ($resql)
+		{
+			$out = '<select class="flat" id="'.$htmlname.'" name="'.$htmlname.'">';
+
+			$num = $this->db->num_rows($resql);
+
+			if ($num > 0)
+			{
+				// Last seen cycle
+				$ref = 0;
+				while ($obj = $this->db->fetch_object($resql))
+				{
 					//Same cycle ?
-			        if ($obj->situation_cycle_ref != $ref) {
+					if ($obj->situation_cycle_ref != $ref)
+					{
 						// Just seen this cycle
-			            $ref = $obj->situation_cycle_ref;
-						//not final ?
-			            if ($obj->situation_final != 1) {
-							//Not prov?
-			                if (substr($obj->facnumber, 1, 4) != 'PROV') {
```

Comment:
Is removing this test on ref of invoice = '(PROV...' done on purpose ?

Reply:
@eldy 
I moved some tests of the "while" loop in the sql query for better performance !

```
$sql = 'SELECT rowid, facnumber, situation_cycle_ref, situation_counter';
$sql.= ' FROM ' . MAIN_DB_PREFIX . 'facture';
$sql.= ' WHERE entity IN ('.getEntity('invoice').')';
$sql.= ' AND situation_counter >= 1';
$sql.= " AND facnumber NOT LIKE '%PROV%'";
$sql.= " AND situation_final != 1";
if (! empty($socid)) {
	$sql.= ' AND fk_soc = ' . $socid;
}
$sql.= ' ORDER BY situation_cycle_ref, situation_counter DESC';
```

Reply:
It is not clear for me what is fixing the bug or not. Also when commiting into already released branch, we must reduce the PR to the lowest change as possible, even if code can be better or optimize, we must try to change the minimal code the fix the bug and only fix the bug to reduce side effect.
Can you make a PR that contains only the bug fix of missing filter on entity ?

Reply:
@eldy 
ok, look this https://github.com/Dolibarr/dolibarr/pull/10870